### ISSUE TITLE
Should Activation/deactivation be executed on every state exit/entry?

### ISSUE THREAD

**Issue by HenningNT** [dotnet-state-machine/stateless#177]
I thought the purpose of the Activation/deactivation was to be run if needed, not on every state change. For instance, maybe when an IoT speaker has its power restored it would start up in the Playing state, but the connection is not there anymore. The OnActive method could then re-connect to the stream it was playing before it lost power.

As it is implemented at the moment it duplicates the entry/exit actions, and I suspect not very useful. I would like to know you think about it.

**Comment by nblumhardt** [comment]
Hi Henning - I think this is by design, if the `OnActivate()` handler isn't called when transitioning into a state, every `OnActivate()` entry ends up being duplicated as an `OnEntry()` handler. The main difference is the additional `OnActivate()` call triggered by `StateMachine.Activate()`.

I haven't used it in anger however, so I may be overlooking the obvious. @yevhen knows this feature best so might have some input here.

**Comment by HenningNT** [comment]
I hope he has some input, as I'm pretty sure the concept involved activation/deactivation when needed (called externally), not on every entry/exit.

**Comment by yevhen** [comment]
Activation is different from state change. State machine can be resurrected (materialized) in some finite state (i.e. initial state) and no OnEntry handlers will be executed, but you still want to be notified that it is now *active*, so you can start a timer for example. The current implementation mirrors UML 2.0 Active State definition (two different handlers for state transition and state activation).

@HenningNT the use-case you are describing might be better modeled with state transitions like `Connected/Disconnected/Playing` and then you can use `OnActivate` handler to play the stream.

**Comment by HenningNT** [comment]
The example is a bit contrived, yes.

But should a state activation/deactivate function be called on every state transition, or just when needed?

**Comment by yevhen** [comment]
Yes, since it is an activation of a new state.

**Comment by HenningNT** [comment]
But... we have entry actions for that?

**Comment by yevhen** [comment]
See https://github.com/dotnet-state-machine/stateless/issues/177#issuecomment-308240430

> State machine can be resurrected (materialized) in some finite state (i.e. initial state) and no OnEntry handlers will be executed, but you still want to be notified that it is now active, so you can start a timer for example.

**Comment by HenningNT** [comment]
Yes, so activation isn't the same as transitioning to a new state, and executing the Entry action. They need to be two different things.

Consider a larger example:
Let's say we are running distributed simulation system. Each computer runs some simulation objects. Each simulation object maybe does Monte Carlo simulations of high energy photon matter interaction, and each object has a list of photon to simulate. When created the simulator object enters Idle, and requests new job data. When it receives job data it goes to Busy and does the calculation for each photon. When it's done it goes to the Sending state, which sends the result to the result gathering server. After successfully sending the result it goes back to Idle again.

But computers has a bad habit of failing, or needing maintenance, so we to deactive the simulation. We want this to happen without screwing up the recorded partial result, or sending the result twice.
On activation we would like the simulation to resume from the point it was at when deactivated.

Something like this:

simulator.Configure(State.Idle)
                .OnEntry(ContactJobServer)
                .OnActivate(ContactJobServer)
                .Permit(Trigger.JobRecieved, State.Busy);

            simulator.Configure(State.Busy)
                .OnEntry(PrepareResultFile)
                .OnActivate(ResumeSimulation)
                .Permit(Trigger.Done, State.Sending)
                .OnDeactivate(StoreProgress)
                .OnExit(MarkAsComplete);

            simulator.Configure(State.Sending)
                .OnEntry(SendResultToServer)
                .OnActivate(CheckIfResultReceived)
                .Permit(Trigger.ResultsSent, State.Idle);

The Idle state has similar handler for OnEntry and OnActivate, since the do the same thing. But the Busy and Sending state must be different. When entering the Busy state the output file is prepared, which is quite different from the OnActivate handler. The OnActivate handler will resume the simulation from where it left off. The progress was stored in a DeActivate handler, which is different from the OnExit handler. 

If the simulator was deactivated while sending the results, it would need to check if the result has been received successfully. The OnActivate handler could then either send the results again, or make a transition to Idle. When entring the Sending state from Busy, the state machine "knows" it hasn't sent the results, so the Entry and Activation has to be different. Same with Exit and Deactivation.

**Comment by HenningNT** [comment]
This stuff still bugs me...

**Comment by HenningNT** [comment]
Please have a look at this commit:
https://github.com/dotnet-state-machine/stateless/commit/90cce62966012b8703ffdce34a55784e45d57fe2
